### PR TITLE
Sprint 1 Release

### LINKED ISSUE #14: Weekly deployment
- Merge all the tickets with status `done`
- Test the functionality
- Fix problems / Create issues

### PR DESCRIPTION
**Pull request by oleksandr-jr**

### CODE REVIEW

**Review by Y-Maxym — APPROVED** [review]
Found a few features:

1. The subreddit DTO does not have a reference to the user, so the subreddit creation endpoint does not work properly. In the future, through security, I think it will be possible to implement so as to extract the id from the currently logged-in user.

2. This is rather not a problem, but the logic of the work that follows from the previous point. If you perform a search by subreddit id, it searches for a subreddit through a join with the user, that is, if you make null in the field (for testing, it was necessary), then it does not find such a subreddit. But if you set optional = true in the Subreddit model above the user field and also remove nullable, then the request will be left join, the result of which will be 100% given by the subreddit, even if the user field = null

File: src/main/java/ua/com/javarush/oleksandr/reddit/redditcloneabstract/controller/PostController.java [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+    private static Post mapPostRequestDtoToPost(PostRequestDTO postRequestDTO) {
+
+        User currentUser = User.with()
+                .userId(postRequestDTO.getUserId())
+                .build();
+
+        return Post.with()
+                .id(postRequestDTO.getId())
+                .user(currentUser)
+                //.subreddit(postRequestDTO.getSubredditName())
```
Comment (aleos-dev) [inline src/main/java/ua/com/javarush/oleksandr/reddit/redditcloneabstract/controller/PostController.java]: We currently use subredditName as an identifier. I propose we rename it to subredditId or guarantee its uniqueness in our model.

File: src/main/java/ua/com/javarush/oleksandr/reddit/redditcloneabstract/controller/PostController.java [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+
+    private static PostResponseDTO mapPostToPostResponseDto(Post post) {
+
+        return PostResponseDTO.with()
+                .id(post.getId())
+                .postName(post.getPostName())
+                .description(post.getDescription())
+                .url(post.getUrl())
+                .voteCount(post.getVoteCount())
+                //.subreddit(post.getSubreddit().getName())
```
Comment (aleos-dev) [inline src/main/java/ua/com/javarush/oleksandr/reddit/redditcloneabstract/controller/PostController.java]: We should uncomment this line.

File: src/main/java/ua/com/javarush/oleksandr/reddit/redditcloneabstract/dto/PostRequestDTO.java [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+    @Length(max = 255, message = "Post name must not exceed 255 characters.")
+    private String postName;
+
+    @Length(max = 255, message = "Description must not exceed 255 characters.")
+    private String description;
+
+    @URL(message = "Bad URL format")
+    private String url;
+
+    //@NotBlank(message = "Subreddit name must not be blank.")
```
Comment (aleos-dev) [inline src/main/java/ua/com/javarush/oleksandr/reddit/redditcloneabstract/dto/PostRequestDTO.java]: Make the necessary corrections and uncomment the code line accordingly to

https://github.com/JR-Abstract/reddit-clone-abstract/pull/39/files#r1507634647

**Comment by oleksandr-jr** [comment]
@aleos-dev  @OnDecember  Requested changes were implemented here #43, #44

File: src/main/java/ua/com/javarush/oleksandr/reddit/redditcloneabstract/model/Post.java [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+
+    @URL
+    @NaturalId
+    @Column(nullable = false)
+    private String url;
+
+    @Column(name = "vote_count")
+    private int voteCount = 0;
+
+    /* @ToString.Exclude
```
Comment (aleos-dev) [inline src/main/java/ua/com/javarush/oleksandr/reddit/redditcloneabstract/model/Post.java]: We should uncomment these lines.

**Review by aleos-dev — APPROVED** [review]
We currently use subredditName as an identifier of Subreddit in PostRequestDTO and PostController. I propose we rename it to subredditId or guarantee SubredditName uniqueness in our Subreddit model. 

The service methods are expected to return consistent data; therefore, it should either return an Optional<E> or a direct instance of E. The issue has been identified in both PostService and SubredditService. 

P.S. I wasn't aware that my previous comment had been successfully posted. Now, I understand how the system works.

**Comment by oleksandr-jr** [comment]
> We currently use subredditName as an identifier of Subreddit in PostRequestDTO and PostController. I propose we rename it to subredditId or guarantee SubredditName uniqueness in our Subreddit model.
> 
> The service methods are expected to return consistent data; therefore, it should either return an Optional or a direct instance of E. The issue has been identified in both PostService and SubredditService.
> 
> P.S. I wasn't aware that my previous comment had been successfully posted. Now, I understand how the system works.

@aleos-dev  Have you seen this PR? https://github.com/JR-Abstract/reddit-clone-abstract/issues/43, https://github.com/JR-Abstract/reddit-clone-abstract/pull/44

**Comment by aleos-dev** [comment]
@oleksandr-jr This took me some time to figure out )))

**Pull request merged**